shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuHashJoin.scala

```diff
@@ -199,20 +200,24 @@ trait GpuHashJoin extends GpuExec with HashJoin {
       }
 
       override def hasNext: Boolean = {
-        while (nextCb.isEmpty && (first || stream.hasNext)) {
+        var may_continue = true
+        while (nextCb.isEmpty && may_continue) {
+          val startTime = System.nanoTime()
           if (stream.hasNext) {
```

Comment:
nit: It would be nice to take the chance to clean up the metrics collection using some of our convenience methods/classes.  I think this makes the code much more readable.

```
withResource(new MetricRange(totalTime)) { _ =>
    val upstreamHasNext = withResource(new MetricRange(streamTime)) { _ =>
        stream.hasNext
    }
    if (upstreamHasNext) {
       ...
```

Reply:
will you merge the current commit first or waiting me to change ti to `withResource`?

Reply:
I would prefer to have it fixed first unless it is blocking you in some way.